### PR TITLE
[RNMobile] Add Featured Banner to Image Block (Android Only)

### PR DESCRIPTION
**Pull request by SiobhyB**
Paritial fix for: wordpress-mobile/gutenberg-mobile#1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3371
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/14451

## Description

This PR introduces a **Featured** banner, which overlays a post's featured image when it's added to an image block. This will make it easier for users to identify which of the images in their posts, if any, are set as featured directly from the editor.

_Note: This PR introduces the **Featured** banner on Android devices only, support for iOS will follow in future iterations._

### Test Case 1: General Flow

To start with, an image block should be updated to reflect whether its image is featured or not when set via the general Post Settings flow:

* Tap the three icons to the upper right of the post editor to access **Post Settings** and set a featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner displays overlays the correct image block.
* Return to **Post Settings**. Tap the existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the image block, as expected.

### Test Case 2: "Editor Loads" Flow

Next, we should verify that a **Featured** banner appears when a post with a featured image is first loaded. To test this, follow these steps:

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it.

### Test Case 3: "Replace in Image Block" Flow

Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the **Featured** banner updates as expected depending on the image that's chosen for an image block:

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
|  |  |

## Types of changes

This PR introduces a new feature (a non-breaking change that adds functionality). Specifically, it adds a **Featured** banner that overlays an image block containing a post's featured image. The code can be split into two overarching categories:

### Badge Component

* A [new Badge component](https://github.com/WordPress/gutenberg/pull/30806/files#diff-f3d1866bb6de6b07c40f6cc1e97d08ef289dc2b4874deec8c414541e0dd5844e) has been created, which adds a "badge" with some text to its child component's upper left. 
* In this case, this new component is used to add a **Featured** badge to an image block. 
* The component was created with re-usability in mind and [the README](https://github.com/WordPress/gutenberg/pull/30806/files#diff-2500f619901448109c703529708b09a157df9072a96e99c19d1d217d54e9aa24) describes how it may be used in other areas of the codebase.

### Logic for Fetching Featured Image ID

* To fetch the ID of a post's initial featured image, [a prop is passed over the bridge](https://github.com/WordPress/gutenberg/pull/30806/files#diff-906aa52029680055db7dc413437ac227ebeaef57c0da27347a019fc53506099a) and set in [the post object](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94R146).
* An [event subscription](https://github.com/WordPress/gutenberg/pull/30806/files#diff-bf4c994cef3006848617e4cd8a8ad4bfbc79ef6feda4d58152369ec92019d51e) is used to track a change to a post's featured image from the Android app. The `editEntityRecord` selector is then used in combination with listeners to [update the value stored for a post's featured image ID](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94). 
* The above enables a `featuredImageId` to be passed to and [used in the image component's edit.native.js file](https://github.com/WordPress/gutenberg/pull/30806/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1). In order to determine whether or not to display the badge component, the `featuredImageId` is compared to the ID of the image in the block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by hypest** [comment]
Thanks for separate PR just to cover the "display featured banner" functionality @SiobhyB , I will take a look tomorrow!

**Comment by hypest** [comment]
Testing results:
✅ Test Case 1: General Flow
✅ Test Case 2: "Editor Loads" Flow
✅ Test Case 3: "Replace in Image Block" Flow

File: packages/block-library/src/image/edit.native.js
Comment (hypest) [inline packages/block-library/src/image/edit.native.js]: Good to see this removal Siobhan! I've been meaning to comment on the PR for it, in the sense that the Image component shouldn't keep `featuredImageId` in its attribute list as it's not an attribute of the block itself but rather, a property of the global post object. 👍

File: packages/block-library/src/image/edit.native.js
Comment (hypest) [inline packages/block-library/src/image/edit.native.js]: Looks like the `androidOnly` const is not used anywhere, can we remove it?

File: packages/components/src/mobile/badge/README.md
Comment (hypest) [inline packages/components/src/mobile/badge/README.md]: > add

Typo perhaps? Replace with "adds"?

File: packages/edit-post/src/editor.native.js
Comment (hypest) [inline packages/edit-post/src/editor.native.js]: I think using the store name directly is not preferred anymore (see [this PR that recently refactored lots of code to switch to using the exposed name](https://github.com/WordPress/gutenberg/pull/27414) instead). Let's follow the same pattern as the `editPostStore` above, thanks!

File: packages/block-library/src/image/edit.native.js
Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js]: Thank you!

File: packages/block-library/src/image/edit.native.js
Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js]: Ah, good point! I tested and there aren't any issues loading posts in iOS with this PR, so the androidOnly tag seems safe to remove. I've removed that now.

File: packages/components/src/mobile/badge/README.md
Comment (SiobhyB) [reply packages/components/src/mobile/badge/README.md]: A typo indeed, thank you for spotting that! I've updated.

File: packages/edit-post/src/editor.native.js
Comment (SiobhyB) [reply packages/edit-post/src/editor.native.js]: Ah, thank you for raising that and the additional context, I've gone ahead to update it. 🙇‍♀️

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt
Comment (hypest) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt]: Kinda a nit but, wdyt about just using `featuredImageId` pretty much everywhere, except the Post object itself where the property is named `featuredMedia`? Both the apps and Gutenberg-desktop refer to "featured image" on the UI and code. I feel like using the `Media` part in places just adds some confusion. Thoughts?

Comment (SiobhyB) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt]: I'd had this on my mind a little too, I used `featuredMedia` as it was a bit more inline with `getEditedPostAttribute('featured_media')`, but agree that using `featuredImageId` seems clearer overall. I've updated this PR and will update the Android PR too when the tests have run.

File: packages/edit-post/src/editor.native.js
Comment (hypest) [inline packages/edit-post/src/editor.native.js:84]: Friendly reminder to resist the urge to have different type of changes in the same commit, unless they form some union of sorts. In this case (commit 427845acbbeb52f071fcf56f8225228106dcb1e8), cleaning the "androidOnly" code and using destructuring for those props are not a union so I'd put them in separate commits.

In many cases, our local file changes probably include all the changes, and what I use `for putting them in separate commits is `git add -p` to interactively select which individual changes to stage. [This article seems to have some good material on this](https://johnkary.net/blog/git-add-p-the-most-powerful-git-feature-youre-not-using-yet/). Other (GUI?) git clients have the feature as well, usually named as simply "staging specific lines".

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt
Comment (hypest) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt]: Should also add that I'm not sure if WordPress would accept a _video_ as featured on a post, in which case "featured media" is a more appropriate name. Gutenberg-desktop doesn't seem to offer videos when selecting a featured media though.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (hypest) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: (Another nit?) Now that on the JS side we use the core-data Store for featured image id storage, I can't quite be sure what we mean when we say "Current". I know that [it was me not long before](https://github.com/WordPress/gutenberg/pull/28854#discussion_r598733588) that proposed the name, but back then I felt that the source of truth is the data we keep in the native side. That's kinda still true but when we'll add the functionality of setting the featured image from the Image block, the source of truth will the be image block itself or the updated core-data store if we go that way. In that case, what's "current"?

Long story short, wdyt about renaming to `featuredImageIdNativeUpdated`? This will also help on the JS side where reading about subscribing to such an event it's clear that it's data held on the native side.

File: packages/edit-post/src/editor.native.js
Comment (SiobhyB) [reply packages/edit-post/src/editor.native.js:84]: Thank you for your insight here and for emphasising the importance of separate, clear commits. I've gone ahead to separate these two out and will bear this in mind for the future.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (SiobhyB) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Your reasoning behind the name change makes sense to me here, I've gone ahead to make that update in this PR and the Android update will follow soon.

**Review by hypest — APPROVED** [review]
The Gutenberg side PR looks good to me at this stage @SiobhyB , good job and thank you for all your follow ups to my review feedback! ❤️

I understand that the WPAndroid side PR is still under review so let's hold on merging the Gutenberg side until all PRs are ready and green.

In the meantime though, may I suggest updating the PR description to match the latest of the code changes? In particular, I'd suggest expanding the "Types of changes" section with details of the various qualitatively different changes in the PR. For example, the introduction of the Badge component, using an event subscription for the featured id change from native and so on. Ideally, each bullet would have some small explanation of the use or need of the change (especially if it's non trivial to understand the "why"), plus a link to the diff or commit so the reader can dig in if they want. An example of such an expanded description could be [this PR I opened recently](https://github.com/WordPress/gutenberg/pull/29683). Let me know if anything is not clear, thanks!

**Comment by hypest** [comment]
Come to think of it again, the bridge changes might actually break the WPiOS app since we're only adding the related support on WPAndroid's side, right? Can you check @SiobhyB ? Thanks!

**Comment by SiobhyB** [comment]
@hypest, thank you for this! A fix has been committed now. ✅ I'm still working on cleaning up commit history for the other two repositories, also. I'll comment on the Android issue when I'm done there.

_Edited to add: I've also added to the PR, as per your guidance. Let me know if there are any further improvements you see. :)_

**Comment by hypest** [comment]
> Edited to add: I've also added to the PR, as per your guidance. Let me know if there are any further improvements you see. :)

Wow, looks awesome to me, thanks! Let me know if you find this part verbose perhaps or not, curious about your thoughts after adding the info.

**Comment by hypest** [comment]
In the meantime, I verified the crash on open (screenshot shared below) and the https://github.com/WordPress/gutenberg/pull/30806/commits/f520659cc578757f9e800b8e308c2a783175a7aa fix! 🎉

**Comment by SiobhyB** [comment]
> Wow, looks awesome to me, thanks! Let me know if you find this part verbose perhaps or not, curious about your thoughts after adding the info.

Thank you!

I liked creating a summary of the changes. It's probably something I'll try to do "as I go" in the future, as it seems it might be a handy "tool" for making sure the code makes sense to me and a good way to spot any room for improvement early on. 

I could have likely been more succinct in this, but I think the wordiness is a product of my newness to code and not having an instinct for the most important parts. I'm sure that'll change given time. :)

**Comment by hypest** [comment]
> I could have likely been more succinct in this, but I think the wordiness is a product of my newness to code and not having an instinct for the most important parts. I'm sure that'll change given time. :)

Just to be clear, it doesn't read verbose to me. But I was interested to hear how did you find the experience of crafting that sections and whether at any point felt too "bureaucratic" or verbose perhaps.

**Review by hypest — APPROVED** [review]
LGTM @SiobhyB !

Here's what I think can be the plan now:
0. I will LGTM all three PRs (Gutenberg, gutenberg-mobile, WPAndroid)
1. I will merge the Gutenberg PR 🎉
2. Please then, update the gutenberg-mobile PR to point to the newly merged hash on Gutenberg's trunk
3. Wait for CI to get green in gutenberg-mobile. Feel free to use the "automatically merge on green" functionality.
4. Please get in touch with the native block editor release wranglers to ensure the two processes (release and this PRs merging) align well.
4. Update the WPAndroid PR to point to the a properly tagged gb-mobile pointer. 
5. After CI is green on WPAndroid, merge it!
6. Celebrate! 😄

**Comment by SiobhyB** [comment]
Thank you @hypest! Just noting here for posterity that, after discussing with the release wranglers, we'll be waiting until next week to merge these changes.

**Comment by SiobhyB** [comment]
@hypest, would you be able to merge this today? I'll then go ahead with the other steps you listed. :) Thank you in advance!

**Comment by hypest** [comment]
> @hypest, would you be able to merge this today? I'll then go ahead with the other steps you listed. :) Thank you in advance!

Yeap, let's get this done! 😸

**Review by hypest — APPROVED** [review]
Still looking good to me, noticed the CHANGELOG.md changes as well and looks like the "Unreleased" section will need some updating to match the current (at the time of writing) native editor release v1.51.0, but that will be wrangled separately so, all good here 👍.

**Pull request merged**
